core/store/src/db.rs

```diff
@@ -79,6 +80,119 @@ impl DBTransaction {
 
 pub type DBIterator<'a> = Box<dyn Iterator<Item = io::Result<(Box<[u8]>, Box<[u8]>)>> + 'a>;
 
+/// Data returned from the database.
+///
+/// Abstraction layer for data returned by [`Database::get_raw_bytes`] and
+/// [`Database::get_with_rc_stripped`] methods.  Operating on the value as
+/// a slice is free while converting it to a vector on an arc may requires an
+/// allocation and memory copy.
+pub struct DBBytes<'a>(DBBytesInner<'a>);
+
+enum DBBytesInner<'a> {
+    /// Data held as a vector.
+    Vec(Vec<u8>),
+
+    /// Data held as RocksDB-specific pinnable slice.
+    Rocks {
+        /// Slice view of the data.
+        ///
+        /// Having this allows us to create views at only parts of the pinnable
+        /// slice without having to modify it.  We need it because remove_suffix
+        /// method of the C++ class is not exposed.
+        slice: &'a [u8],
+
+        /// This is what owns the data.
+        ///
+        /// This is never modified which is why we can have `slice` field
+        /// pointing at the data this pinnable slice holds.
+        db_slice: ::rocksdb::DBPinnableSlice<'a>,
+    },
+}
+
+impl<'a> DBBytes<'a> {
+    pub fn as_slice(&self) -> &[u8] {
+        match self.0 {
+            DBBytesInner::Vec(ref bytes) => bytes.as_slice(),
+            DBBytesInner::Rocks { slice, .. } => slice,
+        }
+    }
+
+    fn from_rocksdb_slice(db_slice: ::rocksdb::DBPinnableSlice<'a>) -> Self {
+        let slice = &*db_slice;
+        let (ptr, len) = (slice.as_ptr(), slice.len());
+        // SAFETY: We got the raw parts from a slice so they are obviously
+        // correct.  By using 'a lifetime we guarantee that this slice won’t
+        // outlive `db_slice` field.  And lastly, since `db_slice` is never
+        // modified and it holds pointers to pinned (in Rust sense) data, we can
+        // safely keep reference to its interior.
+        let slice = unsafe { std::slice::from_raw_parts(ptr, len) };
+        DBBytes(DBBytesInner::Rocks { slice, db_slice })
+    }
+
+    fn strip_refcount(self) -> Option<Self> {
+        match self.0 {
+            DBBytesInner::Vec(bytes) => {
+                refcount::strip_refcount(bytes).map(|bytes| Self(DBBytesInner::Vec(bytes)))
+            }
+            DBBytesInner::Rocks { slice, db_slice } => refcount::decode_value_with_rc(slice)
+                .0
+                .map(|slice| Self(DBBytesInner::Rocks { slice, db_slice })),
+        }
+    }
+}
+
+impl<'a> std::ops::Deref for DBBytes<'a> {
+    type Target = [u8];
+    fn deref(&self) -> &[u8] {
+        self.as_slice()
+    }
+}
+
+impl<'a> From<Vec<u8>> for DBBytes<'a> {
```

Comment:
I'd maybe avoid adding this impl -- in an ideal world, we would only have one variant here, which would always be a slice of data from db. But I can also see that even in the ideal world we'd have an owning version of DBBytes for convenience, so no strong opiniont here.

Reply:
I’m not sure about this either to be honest.  Partially also because I’ve also looked into changing the representation to use Arc rather than a Vec.  The upside of Arc is that we could optimise TestsDB to use Arc to store the data and that would save us copy.  The flip side to that is that we don’t care about optimising TestDB.  And the downside of Arc is that strip_refcount cannot be done zero-copy.  But perhaps converting from Vec should not be a public interface in case we do want to ever change how DBBytes stores data.